### PR TITLE
[narwhal / bullshark] Delete dynamic committee change + epoch change from NW

### PR DESCRIPTION
**Pull request by gdanezis**
The logic to dynamically configure the committee and epoch in narwhal was fragile, and difficult to reason about and maintain. We are moving to an approach where we start separate fresh instances of primaries and workers for epochs or even within epochs. This PR removes the NewEpoch and CommitteeChange messages, the logic they called and the tests that test them.

Bonus: we eliminate the ArcSwaps that were used throughout the code, which are primitives for lock free concurrency, rather than primitives to bypass the &/&mut rules or the need for synchronization, and were used in ways that were hard to reason about and maintain.

We keep the logic to cleanly shut down an instance and its workers.

### CODE REVIEW

**Review by andll — APPROVED** [review]
This is epic! This `committee.load()` and online worker reconfiguration was indeed scary part of the code.

File: narwhal/node/src/restarter.rs [hunk 1]
```diff
@@ -130,7 +129,7 @@ impl NodeRestarter {
             name = primary_keypair.public().clone();
             worker_ids_and_keypairs = new_worker_ids_and_keypairs;
             committee = new_committee;
-            worker_cache.swap(Arc::new(new_worker_cache));
+            // worker_cache.swap(Arc::new(new_worker_cache));
```
Comment (bmwill) [inline narwhal/node/src/restarter.rs:132]: remove?

**Review by akichidis — CHANGES_REQUESTED** [review]
Haven't reviewed the PR yet - so this is not a blocker to change something :-) - but as I am still evaluating the reconfiguration path I would like to hold and not merge until confirmed the plan. @gdanezis thanks for the PR!

**Review by bmwill — APPROVED** [review]

File: narwhal/node/src/lib.rs [hunk 2]
```diff
@@ -61,8 +61,8 @@ impl Node {
     where
         State: ExecutionState + Send + Sync + 'static,
     {
-        let initial_committee = ReconfigureNotification::NewEpoch((**committee.load()).clone());
-        let (tx_reconfigure, _rx_reconfigure) = watch::channel(initial_committee);
+        let initial_committee = ShutdownNotification::Run;
+        let (tx_shutdown, _rx_shutdown) = watch::channel(initial_committee);
```
Comment (laura-makdah) [inline narwhal/node/src/lib.rs:65]: will there be a need to read the committee from the channel when it is now only being used as shutdown?

File: narwhal/node/src/lib.rs [hunk 3]
```diff
@@ -218,16 +218,16 @@ impl Node {
...
             parameters.gc_depth,
             consensus_metrics.clone(),
         );
         let consensus_handles = Consensus::spawn(
-            (**committee.load()).clone(),
+            (*committee).clone(),
             store.consensus_store.clone(),
             store.certificate_store.clone(),
-            tx_reconfigure.subscribe(),
+            tx_shutdown.subscribe(),
```
Comment (laura-makdah) [inline narwhal/node/src/lib.rs:230]: nit: let's also unify our approach to either pass receivers, or senders which are then used to subscribe in the spawned component (this is the only outlier?)

**Review by laura-makdah — COMMENTED** [review]
I don't see any major issues. Thanks!

**Review by asonnino — APPROVED** [review]

File: narwhal/node/tests/reconfigure.rs [hunk 4]
```diff
@@ -1,379 +0,0 @@
...
-                tx = TransactionProto {
-                    transaction: Bytes::from(epoch.to_le_bytes().to_vec()),
-                };
-            }
-        }
-    }
-}
-
-#[tokio::test]
-async fn restart() {
```
Comment (akichidis) [inline narwhal/node/tests/reconfigure.rs:166]: Let's keep this test, it is needed as it is testing the reconfiguration approach via the `restart/shutdown` method. I'll rebase the PR as I've made some changes on this test

File: narwhal/primary/src/grpc_server/configuration.rs [hunk 5]
```diff
@@ -84,43 +83,8 @@ impl Configuration for NarwhalConfiguration {
     #[allow(clippy::mutable_key_type)]
     async fn new_network_info(
         &self,
-        request: Request<NewNetworkInfoRequest>,
```
Comment (akichidis) [inline narwhal/primary/src/grpc_server/configuration.rs:87]: As far as I remember this was a request from Celo/Sommelier. The idea is that networking information might not be available from the beginning and change down the line. It seems that with this PR we are removing completely this ability. Apart from that, I am not convinced that this endpoint was working correctly even before either. It seems that we do swap the committee , but never update the actual networking info via anemo. That being said I would at least add here a TODO to indicate a breaking behaviour

**Pull request closed without merging**